File: GTAClone-iOS/GTAClone/ui/HUDView.swift
import UIKit

class HUDView: UIView {

    private let world: World3D
    private let joystick: VirtualJoystick

    init(frame: CGRect, world: World3D, joystick: VirtualJoystick) {
        self.world = world; self.joystick = joystick
        super.init(frame: frame)
        backgroundColor = .clear; isOpaque = false
        joystick.configure(cx: frame.width * 0.13, cy: frame.height * 0.72, radius: frame.height * 0.20)
    }
    required init?(coder: NSCoder) { fatalError() }

    override func draw(_ rect: CGRect) {
        guard let ctx = UIGraphicsGetCurrentContext() else { return }
        let w = rect.width; let h = rect.height
        joystick.draw(in: ctx)
        drawMiniMap(ctx, w: w, h: h)
        drawHealth(ctx, w: w, h: h)
        drawMoney(ctx, w: w)
        drawWanted(ctx, w: w)
        drawClock(ctx, w: w)
        drawButtons(ctx, w: w, h: h)
        if world.player.isInVehicle { drawHint(ctx, w: w, h: h) }
    }

    // ── Mini-map ──────────────────────────────────────────────────────────────
    private func drawMiniMap(_ ctx: CGContext, w: CGFloat, h: CGFloat) {
        let size = h * 0.22; let mx: CGFloat = 20; let my: CGFloat = 20
        let cs = CGFloat(world.city.citySize()); let scale = size / cs

        // Background
        ctx.setFillColor(UIColor(red:0.1,green:0.16,blue:0.1,alpha:0.75).cgColor)
        UIBezierPath(roundedRect: CGRect(x:mx,y:my,width:size,height:size), cornerRadius: 10).fill()

        // Road marks
        ctx.setFillColor(UIColor(white: 0.33, alpha: 1).cgColor)
        for slab in world.city.city.roadSlabs {
            // roadSlabs are Box3D from City3D — map center/size → minimap rect
        }
        // Simplified: draw road grid lines
        ctx.setStrokeColor(UIColor(white: 0.4, alpha: 1).cgColor); ctx.setLineWidth(1)
        let interval = CGFloat(world.city.BLOCK + world.city.ROAD) * scale
        var pos = CGFloat(world.city.BLOCK) * scale
        while pos < size { ctx.move(to: .init(x: mx+pos, y: my)); ctx.addLine(to: .init(x: mx+pos, y: my+size)); pos += interval }
        pos = CGFloat(world.city.BLOCK) * scale
        while pos < size { ctx.move(to: .init(x: mx, y: my+pos)); ctx.addLine(to: .init(x: mx+size, y: my+pos)); pos += interval }
        ctx.strokePath()

        // NPCs
        ctx.setFillColor(UIColor.white.cgColor)
        for n in world.npcs where n.health > 0 { ctx.fillEllipse(in: dot(mx + CGFloat(n.x)*scale, my + CGFloat(n.z)*scale, 2)) }

        // Vehicles
        ctx.setFillColor(UIColor.yellow.cgColor)
        for v in world.vehicles { ctx.fillEllipse(in: dot(mx + CGFloat(v.x)*scale, my + CGFloat(v.z)*scale, 3)) }

        // Player
        ctx.setFillColor(UIColor.green.cgColor)
        ctx.fillEllipse(in: dot(mx + CGFloat(world.player.x)*scale, my + CGFloat(world.player.z)*scale, 5))

        ctx.setStrokeColor(UIColor(white: 1, alpha: 0.65).cgColor); ctx.setLineWidth(2)
        UIBezierPath(roundedRect: CGRect(x:mx,y:my,width:size,height:size), cornerRadius: 10).stroke()
    }

    private func dot(_ x: CGFloat, _ y: CGFloat, _ r: CGFloat) -> CGRect { CGRect(x:x-r,y:y-r,width:r*2,height:r*2) }

    // ── Health ────────────────────────────────────────────────────────────────
    private func drawHealth(_ ctx: CGContext, w: CGFloat, h: CGFloat) {
        let bx: CGFloat = 20; let by = h*0.22 + 32; let bw: CGFloat = 180; let bh: CGFloat = 18
        let ratio = CGFloat(world.player.health / 100).clamped(0, 1)
        ctx.setFillColor(UIColor.black.withAlphaComponent(0.5).cgColor)
        UIBezierPath(roundedRect: CGRect(x:bx-4,y:by-4,width:bw+60,height:bh+8), cornerRadius: 5).fill()
        let hpColor: UIColor = ratio > 0.6 ? UIColor(red:0.3,green:0.76,blue:0.3,alpha:1) :
                               ratio > 0.3 ? UIColor(red:1,green:0.75,blue:0.1,alpha:1) :
                                             UIColor(red:0.95,green:0.25,blue:0.25,alpha:1)
        ctx.setFillColor(hpColor.cgColor)
        if ratio > 0 { UIBezierPath(roundedRect: CGRect(x:bx,y:by,width:bw*ratio,height:bh), cornerRadius: 4).fill() }
        let attrs: [NSAttributedString.Key: Any] = [.foregroundColor: UIColor.white, .font: UIFont.monospacedSystemFont(ofSize: 14, weight: .regular)]
        "HP \(Int(world.player.health))".draw(at: CGPoint(x: bx+bw+8, y: by+1), withAttributes: attrs)
    }

    // ── Money ─────────────────────────────────────────────────────────────────
    private func drawMoney(_ ctx: CGContext, w: CGFloat) {
        let s = String(format: "$%08d", world.player.money)
        let attrs: [NSAttributedString.Key: Any] = [.foregroundColor: UIColor(red:0.2,green:1,blue:0.2,alpha:1),
                                                     .font: UIFont.monospacedSystemFont(ofSize: 32, weight: .bold)]
        let sz = (s as NSString).size(withAttributes: attrs)
        s.draw(at: CGPoint(x: w - sz.width - 20, y: 18), withAttributes: attrs)
    }

    // ── Wanted stars ──────────────────────────────────────────────────────────
    private func drawWanted(_ ctx: CGContext, w: CGFloat) {
        let wl = world.player.wantedLevel
        for i in 0..<5 {
            let col: UIColor = i < wl ? .yellow : UIColor(white: 0.25, alpha: 1)
            let attrs: [NSAttributedString.Key: Any] = [.foregroundColor: col, .font: UIFont.systemFont(ofSize: 24, weight: .bold)]
            "★".draw(at: CGPoint(x: w - 180 + CGFloat(i)*34, y: 60), withAttributes: attrs)
        }
    }

    // ── Clock ─────────────────────────────────────────────────────────────────
    private func drawClock(_ ctx: CGContext, w: CGFloat) {
        let s = "12:45"
        let attrs: [NSAttributedString.Key: Any] = [.foregroundColor: UIColor.white,
                                                     .font: UIFont.monospacedSystemFont(ofSize: 26, weight: .bold)]
        let sz = (s as NSString).size(withAttributes: attrs)
        ctx.setFillColor(UIColor.black.withAlphaComponent(0.5).cgColor)
        UIBezierPath(roundedRect: CGRect(x: w/2-sz.width/2-12, y: 10, width: sz.width+24, height: 40), cornerRadius: 8).fill()
        s.draw(at: CGPoint(x: w/2 - sz.width/2, y: 16), withAttributes: attrs)
    }

    // ── Action buttons ────────────────────────────────────────────────────────
    private func drawButtons(_ ctx: CGContext, w: CGFloat, h: CGFloat) {
        let bs = h * 0.11; let mg: CGFloat = 20
        let rx = w - mg - bs; let by = h - mg - bs
        drawBtn(ctx, x: rx - bs*1.3, y: by - bs*1.3, s: bs, color: UIColor(red:1,green:0.67,blue:0,alpha:0.55), label: "CAR")
        drawBtn(ctx, x: rx, y: by, s: bs, color: UIColor(red:1,green:0.27,blue:0.27,alpha:0.55), label: "ATK")
        let runColor = world.player.sprinting ? UIColor(red:0.27,green:0.67,blue:1,alpha:0.80) : UIColor(red:0.27,green:0.67,blue:1,alpha:0.33)
        drawBtn(ctx, x: rx, y: by - bs*1.3, s: bs, color: runColor, label: "RUN")
    }

    private func drawBtn(_ ctx: CGContext, x: CGFloat, y: CGFloat, s: CGFloat, color: UIColor, label: String) {
        ctx.setFillColor(color.cgColor)
        UIBezierPath(roundedRect: CGRect(x:x,y:y,width:s,height:s), cornerRadius: s*0.28).fill()
        ctx.setStrokeColor(UIColor(white:1,alpha:0.5).cgColor); ctx.setLineWidth(2)
        UIBezierPath(roundedRect: CGRect(x:x,y:y,width:s,height:s), cornerRadius: s*0.28).stroke()
        let attrs: [NSAttributedString.Key: Any] = [.foregroundColor: UIColor.white,
                                                     .font: UIFont.boldSystemFont(ofSize: s * 0.27)]
        let sz = (label as NSString).size(withAttributes: attrs)
        label.draw(at: CGPoint(x: x + (s-sz.width)/2, y: y + (s-sz.height)/2 + 2), withAttributes: attrs)
    }

    // ── Vehicle hint ──────────────────────────────────────────────────────────
    private func drawHint(_ ctx: CGContext, w: CGFloat, h: CGFloat) {
        let msg = "DRAG RIGHT TO ORBIT  ·  TAP CAR TO EXIT"
        let attrs: [NSAttributedString.Key: Any] = [.foregroundColor: UIColor(red:1,green:0.67,blue:0,alpha:1),
                                                     .font: UIFont.monospacedSystemFont(ofSize: 16, weight: .regular)]
        let sz = (msg as NSString).size(withAttributes: attrs)
        ctx.setFillColor(UIColor.black.withAlphaComponent(0.5).cgColor)
        UIBezierPath(roundedRect: CGRect(x: w/2-sz.width/2-14, y: h-48, width: sz.width+28, height: 34), cornerRadius: 8).fill()
        msg.draw(at: CGPoint(x: w/2 - sz.width/2, y: h - 42), withAttributes: attrs)
    }
}

private extension CGFloat {
    func clamped(_ lo: CGFloat, _ hi: CGFloat) -> CGFloat { max(lo, min(hi, self)) }
}

File: app/src/main/kotlin/com/gtaclone/ui/HUDView.kt
package com.gtaclone.ui

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Typeface
import android.view.MotionEvent
import android.view.View
import com.gtaclone.engine.InputManager
import com.gtaclone.world.World3D

class HUDView(context: Context, private val world: World3D, private val joystick: VirtualJoystick) : View(context) {

    private val input = InputManager()

    private val bgPaint  = Paint().apply { isAntiAlias = true }
    private val fillPaint = Paint().apply { isAntiAlias = true }
    private val borderPaint = Paint().apply {
        color = Color.parseColor("#88FFFFFF"); style = Paint.Style.STROKE; strokeWidth = 2f; isAntiAlias = true
    }
    private val moneyPaint = Paint().apply {
        color = Color.parseColor("#00FF00"); textSize = 34f; textAlign = Paint.Align.RIGHT
        typeface = Typeface.create(Typeface.MONOSPACE, Typeface.BOLD); isAntiAlias = true
    }
    private val textPaint = Paint().apply {
        color = Color.WHITE; textSize = 26f; typeface = Typeface.MONOSPACE; isAntiAlias = true
    }
    private val timePaint = Paint().apply {
        color = Color.WHITE; textSize = 28f; textAlign = Paint.Align.CENTER
        typeface = Typeface.create(Typeface.MONOSPACE, Typeface.BOLD); isAntiAlias = true
    }
    private val starPaint = Paint().apply {
        textSize = 26f; typeface = Typeface.DEFAULT_BOLD; isAntiAlias = true
    }
    private val btnText = Paint().apply {
        color = Color.WHITE; textAlign = Paint.Align.CENTER
        typeface = Typeface.DEFAULT_BOLD; isAntiAlias = true
    }
    private val miniPaint = Paint().apply { isAntiAlias = false }

    init { setBackgroundColor(Color.TRANSPARENT) }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        joystick.init(w * 0.13f, h * 0.72f, h * 0.2f)
        invalidate()
    }

    override fun onTouchEvent(ev: MotionEvent): Boolean {
        input.handleTouch(ev, joystick, world, width.toFloat(), height.toFloat())
        return true
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        joystick.draw(canvas)
        drawMiniMap(canvas)
        drawHealth(canvas)
        drawMoney(canvas)
        drawWanted(canvas)
        drawTime(canvas)
        drawButtons(canvas)
        if (world.player.isInVehicle) drawVehicleHint(canvas)
        postInvalidateOnAnimation()
    }

    // ── Mini-map ──────────────────────────────────────────────────────────────
    private fun drawMiniMap(canvas: Canvas) {
        val size = height * 0.22f
        val mx = 20f; val my = 20f
        val city = world.city; val cs = city.citySize()
        val scale = size / cs

        bgPaint.color = Color.parseColor("#BB1a2a1a")
        canvas.drawRoundRect(mx, my, mx + size, my + size, 10f, 10f, bgPaint)

        // Road grid dots
        miniPaint.color = Color.parseColor("#555555")
        city.roadSlabs.forEach { s ->
            val px = mx + s.cx * scale; val py = my + s.cz * scale
            canvas.drawRect(px - s.w * scale * 0.5f, py - s.d * scale * 0.5f,
                px + s.w * scale * 0.5f, py + s.d * scale * 0.5f, miniPaint)
        }

        // NPCs
        miniPaint.color = Color.WHITE
        world.npcs.filter { it.health > 0f }.forEach { n ->
            canvas.drawCircle(mx + n.x * scale, my + n.z * scale, 2f, miniPaint)
        }
        // Vehicles
        miniPaint.color = Color.YELLOW
        world.vehicles.forEach { v ->
            canvas.drawCircle(mx + v.x * scale, my + v.z * scale, 3f, miniPaint)
        }
        // Player (green)
        miniPaint.color = Color.GREEN
        canvas.drawCircle(mx + world.player.x * scale, my + world.player.z * scale, 5f, miniPaint)

        borderPaint.color = Color.parseColor("#AAFFFFFF")
        canvas.drawRoundRect(mx, my, mx + size, my + size, 10f, 10f, borderPaint)
    }

    // ── Health ────────────────────────────────────────────────────────────────
    private fun drawHealth(canvas: Canvas) {
        val bx = 20f; val by = height * 0.22f + 30f; val bw = 180f; val bh = 18f
        val ratio = (world.player.health / 100f).coerceIn(0f, 1f)
        bgPaint.color = Color.parseColor("#88000000")
        canvas.drawRoundRect(bx - 4f, by - 4f, bx + bw + 50f, by + bh + 4f, 5f, 5f, bgPaint)
        fillPaint.color = when { ratio > 0.6f -> 0xFF4CAF50.toInt(); ratio > 0.3f -> 0xFFFFC107.toInt(); else -> 0xFFF44336.toInt() }
        if (ratio > 0f) canvas.drawRoundRect(bx, by, bx + bw * ratio, by + bh, 4f, 4f, fillPaint)
        textPaint.textSize = 16f
        canvas.drawText("HP ${world.player.health.toInt()}", bx + bw + 6f, by + bh - 1f, textPaint)
    }

    // ── Money ─────────────────────────────────────────────────────────────────
    private fun drawMoney(canvas: Canvas) {
        canvas.drawText("\$${String.format("%08d", world.player.money)}", width - 20f, 54f, moneyPaint)
    }

    // ── Wanted stars ──────────────────────────────────────────────────────────
    private fun drawWanted(canvas: Canvas) {
        val baseX = width - 178f
        starPaint.textSize = 26f; starPaint.textAlign = Paint.Align.LEFT
        for (i in 0 until 5) {
            starPaint.color = if (i < world.player.wantedLevel) Color.YELLOW else Color.parseColor("#444444")
            canvas.drawText("★", baseX + i * 32f, 92f, starPaint)
        }
    }

    // ── Clock ─────────────────────────────────────────────────────────────────
    private fun drawTime(canvas: Canvas) {
        val tw = timePaint.measureText("12:45")
        bgPaint.color = Color.parseColor("#88000000")
        canvas.drawRoundRect(width / 2f - tw / 2 - 12f, 10f, width / 2f + tw / 2 + 12f, 48f, 8f, 8f, bgPaint)
        canvas.drawText("12:45", width / 2f, 42f, timePaint)
    }

    // ── Action buttons ────────────────────────────────────────────────────────
    private fun drawButtons(canvas: Canvas) {
        val bs = height * 0.11f; val mg = 20f
        val rx = width - mg - bs; val by = height - mg - bs
        btn(canvas, rx - bs * 1.3f, by - bs * 1.3f, bs, Color.parseColor("#88FFAA00"), "CAR")
        btn(canvas, rx, by, bs, Color.parseColor("#88FF4444"), "ATK")
        btn(canvas, rx, by - bs * 1.3f, bs,
            if (world.player.sprinting) Color.parseColor("#CC44AAFF") else Color.parseColor("#5544AAFF"), "RUN")
    }

    private fun btn(canvas: Canvas, x: Float, y: Float, s: Float, color: Int, label: String) {
        fillPaint.color = color
        canvas.drawRoundRect(x, y, x + s, y + s, s * 0.28f, s * 0.28f, fillPaint)
        canvas.drawRoundRect(x, y, x + s, y + s, s * 0.28f, s * 0.28f, borderPaint)
        btnText.textSize = s * 0.28f
        canvas.drawText(label, x + s / 2, y + s * 0.62f, btnText)
    }

    private fun drawVehicleHint(canvas: Canvas) {
        textPaint.textAlign = Paint.Align.CENTER; textPaint.textSize = 20f; textPaint.color = Color.parseColor("#FFAA00")
        bgPaint.color = Color.parseColor("#88000000")
        canvas.drawRoundRect(width / 2f - 145f, height - 50f, width / 2f + 145f, height - 14f, 8f, 8f, bgPaint)
        canvas.drawText("DRAG RIGHT SIDE TO ROTATE CAMERA  ·  TAP CAR TO EXIT", width / 2f, height - 22f, textPaint)
    }
}
